File: Lokesh_Saana/47-permutations-ii.py
# https://leetcode.com/problems/permutations-ii/

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        count = {i: 0 for i in nums}
        for i in nums:
            count[i] += 1
        res = []
        perm = []

        def backtrack():
            if len(perm) == len(nums):
                res.append(perm.copy())
                return

            for i in count:
                if count[i] > 0:
                    perm.append(i)
                    count[i] -= 1

                    backtrack()

                    count[i] += 1
                    perm.pop()
        backtrack()
        return res

File: Ayush-Gupta/47. Permutations II.cpp
class Solution {
public:
    void fun(vector<int> nums, int i, int j, vector<vector<int>> &ans) {
        if(i==j) {
            ans.push_back(nums);
            return;
        }
        for(int k=i;k<=j;k++) {
            if(i!=k && nums[i]==nums[k]) continue;
            swap(nums[i], nums[k]);
            fun(nums, i+1, j, ans);
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        fun(nums, 0, nums.size()-1, ans);
        return ans;
    }
};